src/node/transaction.cpp

```diff
@@ -52,27 +53,27 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
             CallFunctionInValidationInterfaceQueue([&promise] {
                 promise.set_value();
             });
+            callback_set = true;
         }
     } else if (fHaveChain) {
         return TransactionError::ALREADY_IN_CHAIN;
-    } else {
-        // Make sure we don't block forever if re-sending
-        // a transaction already in mempool.
-        promise.set_value();
     }
 
     } // cs_main
 
-    promise.get_future().wait();
-
-    if (!g_connman) {
-        return TransactionError::P2P_DISABLED;
+    if (callback_set) {
+        promise.get_future().wait();
     }
 
-    CInv inv(MSG_TX, hashTx);
-    g_connman->ForEachNode([&inv](CNode* pnode) {
-        pnode->PushInventory(inv);
-    });
+    if (relay) {
+        if (!g_connman) {
+            return TransactionError::P2P_DISABLED;
+        }
+        CInv inv(MSG_TX, hashTx);
+        g_connman->ForEachNode([&inv](CNode* pnode) {
```

Comment:
87cca293d5b127e43e35d866167893454658eacd

just noting that this must be changed in #16452 to `RelayTransaction`.